File: 0045-jump-game-ii/0045-jump-game-ii.cpp
int memo[100001];

class Solution {
public:
    int dp(vector<int> &nums, int i) {
        if (i >= nums.size() - 1) {
            return 0;
        }
        if (memo[i] != -1) {
            return memo[i];
        }
        int ans = 1e9;
        for (int j = 1; j <= nums[i]; ++j) {
            ans = min(ans, 1 + dp(nums, i + j));
        }
        return memo[i] = ans;
    }
    int jump(vector<int>& nums) {
        memset(memo, -1, sizeof(memo));
        return dp(nums, 0);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length - 1, l = 0, r = 0, res = 0;
        while (r < n) {
            int f = 0;
            for (int i = l; i < r + 1; i++) {
                f = Math.max(f, i + nums[i]);
            }
            l = r + 1;
            r = f;
            res++;
        }
        return res;
    }
}